platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -17,68 +19,374 @@ components within the platform.
+roster will be the round number of the state + 1. The adoption of a new active roster updates the old roster history
+by causing the old current roster to become the new previous roster. The old previous roster is removed from the
+history and the merkle state. After the new roster history is determined, it is handed to the platform during
+construction.
+
+The `Roster History` is needed by the platform at startup for event validation during Pre-Consensus Event Stream (PCES)
+replay. In the future, the birth round of the event will be used to determine which roster to use for validation. In
+the current implementation, the platform expects two `AddressBooks`, the `current` and `previous` address books.
+The `previous` address book is used to validate events that have older software versions than the current software
+version. When we switch to rosters, the previous and current roster will be used in place of the previous and
+current address book. The previous roster will continue to be used to validate events that have an older software
+version. The switch to using event birth rounds will happen in a future proposal for Dynamic Address Book phase 3.
+
+During the handling of a freeze upgrade transaction, instead of writing the new config.txt to disk, a new candidate
+roster is created and stored in the roster merkle state.
+
+NOTE: The roster state maintains its own `Roster History`. Setting the active roster in the roster state will cause
+the history to be updated by rotating the current active roster to the previous active roster and setting the new
+roster as the current active roster. The old previous active roster is removed from the history and the merkle state.
+
+![StartupAndRosters-TSS-Roster.drawio.svg](StartupAndRosters-TSS-Roster.drawio.svg)
+
+#### Startup: Determining The New Roster History
+
+There are four distinct modes that a node can start in:
+
+* `Genesis Network` - The node is started with a genesis roster and no pre-existing state on disk.
+* `Network Transplant` - The node is started with a state on disk and an overriding roster for a different network.
+* `Software Upgrade` - The node is restarted with the same state on disk and a software upgrade is happening.
+* `Normal Restart` - The node is restarted with the same state on disk and no software upgrade is happening.
+
+A decision tree for deciding which mode to start the network in is as follows:
+
+![](TSS%20Roster%20Lifecycle-Proposed%20Startup%20Behavior.drawio.svg)
+
+##### MODE: New Network Genesis
+
+Conditions:
+
+1. No state loaded from disk
+2. A `genesis-config.txt` file is present on disk
+
+Ignored Files: (Log warning if present)
+
+1. `config.txt`
+2. `override-config.txt`
+
+Startup Tasks:
+
+1. Translate `genesis-config.txt` to a genesisRoster.
+2. Create a genesis state with the genesisRoster.
+3. Set (genesisRoster, 0) ase the new active roster in the roster state.
+4. rosterHistory := `[(genesisRoster, 0)]`
```

Comment:
I wonder if the following statement would simplify the implementation.

```
rosterHistory := [(genesisRoster, 0), (genesisRoster, 0)]
```

Reply:
Probably eliminates any special casing to have both instantiated, sure.